CMake 3.25 and .gitattributes

digitalcarp opened the pull request:
## CMake 3.25

* Bump minimum version to 3.25
* Add CMake presets
* Use PkgConfig `IMPORTED_TARGET` to simplify library integration
* Add `fmt` formatting library using FetchContent
* Modernize CMake usage to prefer target based functions rather than globals
* Refactor CMake scripts

### Example Usage

```bash
cmake --preset dev
cmake --build --preset dev --target install
```

#### User Presets

In the project root directory, create `CMakeUserPresets.json`.

```json
{
    "version": 6,
    "configurePresets": [
        {
            "name": "dev-libraw",
            "inherits": "dev",
            "cacheVariables": {
                "WITH_SYSTEM_LIBRAW": "ON"
            }
        },
        {
            "name": "dev-asan",
            "inherits": "dev",
            "cacheVariables": {
                "WITH_SAN": "address"
            }
        }
    ],
    "buildPresets": [
        {
            "name": "dev-libraw",
            "configurePreset": "dev-libraw"
        },
        {
            "name": "dev-asan",
            "configurePreset": "dev-asan"
        }
    ]
}
```

```bash
cmake --preset dev-asan
cmake --build --preset dev-asan --target install
```

## .gitattributes

Normalize line ending settings for anyone that commits to the repo. For existing repos (with changes),

```bash
git rm --cached -r .
git add --renormalize -A
# If add does not work, either use dos2unix or:
git reset --hard
```

digitalcarp commented on clean.bat:
Ignore: Line ending changes only

digitalcarp commented on clean.bat:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/LibRaw.pro:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/RawSpeed/rawspeed.win32-dll.patch:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/RawSpeed3/README.md:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/buildfiles/4channels.pro:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/buildfiles/dcraw_emu.pro:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/buildfiles/dcraw_half.pro:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/buildfiles/half_mt.pro:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/buildfiles/libraw.pro:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/buildfiles/mem_image.pro:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/buildfiles/multirender_test.pro:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/buildfiles/postprocessing_benchmark.pro:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/buildfiles/raw-identify.pro:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/buildfiles/simple_dcraw.pro:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/buildfiles/unprocessed_raw.pro:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/doc/API-datastruct.html:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/doc/Install-LibRaw.html:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/internal/libraw_cameraids.h:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/src/decoders/crx.cpp:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/src/decoders/fuji_compressed.cpp:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/src/decoders/unpack_thumb.cpp:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/src/demosaic/dcb_demosaic.cpp:
Ignore: Line ending changes only

digitalcarp commented on rtengine/libraw/src/metadata/hasselblad_model.cpp:
Ignore: Line ending changes only

digitalcarp commented on tools/buildRT.bat:
Ignore: Line ending changes only

Lawrence37 replied:
This is a Windows file. Should the Windows line endings be kept instead of converting them to Unix?

digitalcarp commented on win.cmake:
Ignore: Line ending changes only

Lawrence37 commented:
How come the AppImage workflow succeeds if it using CMake 3.22? I thought `cmake_minimum_required` is supposed to throw an error if the minimum version is not satisfied.

digitalcarp commented:
https://github.com/actions/runner-images/blob/ubuntu22/20250330.1/images/ubuntu/Ubuntu2204-Readme.md#tools

The AppImage is using the latest CMake 4.0.0

Lawrence37 commented:
So it uses a different version of CMake despite installing 3.22.1? No wonder why workflows sometimes start failing.

digitalcarp commented:
Looking at our appimage.yml, I only see cmake being listed as a dependency to install and not a specific version. I assume it picks up whatever existing version is bundled with the app image then.

Lawrence37 commented:
It installs the current version in the Ubuntu repository for that release (22.04):
```
Setting up cmake (3.22.1-1ubuntu1.22.04.2) ...
```

digitalcarp commented:
Lowered to 3.15. Tested with: https://cmake.org/files/v3.15/

3.14 wasn't working with dev version of `LibRaw.cmake`.

Lawrence37 commented on CMakeLists.txt:
```diff
@@ -55,43 +47,42 @@ elseif(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSI
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -ffp-contract=off")
 endif()
 
-# We might want to build using the old C++ ABI, even when using a new GCC
-# version:
-if(USE_OLD_CXX_ABI)
-    add_definitions(-D_GLIBCXX_USE_CXX11_ABI=0)
-endif()
```
`USE_OLD_CXX_ABI` is no longer an option?

digitalcarp replied:
It shouldn't be needed nowadays. libexiv2 and gtkmm are the only libraries that use `std::string`. Latest versions of the library are using C++20/C++17 respectively so they aren't depending on old ABI. Neither provide a dedicated flag in their CMake or autoconf build flow to specify using the old ABI either.

If someone is really using libraries built with a 10 year old compiler (gcc 4.x) and still need to build all their binaries using the old ABI (to ensure compatibility across all their programs), they can use `-DCMAKE_CXX_COMPILE_DEFINITIONS="-D_GLIBCXX_USE_CXX11_ABI=0"`. I would expect someone who is still dependent on such a flag to know how to apply it themselves.

I think this option was introduced around 2015 when the ABI break happened. It really shouldn't be an issue even for LTS operating systems.

Lawrence37 replied:
Ok, I'll make a note to update the documentation and add the details in the release notes.

Lawrence37 commented on CMakeLists.txt:
```diff
@@ -443,386 +416,30 @@ if(NOT BUILD_BUNDLE)
-endif()
-
-if(WITH_LTO)
-    # Using LTO with older versions of binutils requires setting extra flags
```
Is this being handled elsewhere?

digitalcarp replied:
All LTO configuration is handled by CMake natively in `cmake/Compile.cmake` via the line:

```cmake
    if(WITH_LTO)
        set_target_properties(${CXX_TARGET} PROPERTIES INTERPROCEDURAL_OPTIMIZATION ON)
    endif()
```

Lawrence37 commented on CMakeLists.txt:
```diff
@@ -443,386 +416,30 @@ if(NOT BUILD_BUNDLE)
-pkg_check_modules(LENSFUN REQUIRED lensfun>=0.2)
-pkg_check_modules(RSVG REQUIRED librsvg-2.0>=2.52)
-
-# Require exiv2 >= 0.24 to make sure everything we need is available
-#find_package(Exiv2 0.24 REQUIRED)
-pkg_check_modules(EXIV2 REQUIRED exiv2>=0.24)
-#include_directories(SYSTEM ${Exiv2_INCLUDE_DIRS})
-#list(APPEND LIBS ${EXIV2_LIBRARIES})
-add_definitions(${EXIV2_DEFINITIONS})
-set(_exiv2_libs ${EXIV2_LIBRARIES})
-set(EXIV2_LIBRARIES "")
-foreach(l ${_exiv2_libs})
-    set(_el "_el-NOTFOUND")
-    if(EXIV2_LIBRARY_DIRS)
-        message(STATUS "searching for library ${l} in ${EXIV2_LIBRARY_DIRS}")
-        find_library(_el ${l} PATHS ${EXIV2_LIBRARY_DIRS} NO_DEFAULT_PATH)
-        if(NOT _el)
-            message(STATUS  "  NOT FOUND, searching again in default path")
-            find_library(_el ${l})
-        endif()
-    else()
-        message(STATUS "searching for library ${l}")
-        find_library(_el ${l})# PATHS ${EXIV2_LIBRARY_DIRS})
-    endif()
-    message(STATUS "  result: ${_el}")
-    set(EXIV2_LIBRARIES ${EXIV2_LIBRARIES} ${_el})
-endforeach()
```
I'm not sure what this complex Exiv2 logic is for, but do you know for sure that it is not needed anymore?

digitalcarp replied:
It's searching for the exiv2 libraries. If PkgConfig can't find the library files, the build should fail due to the `REQUIRED` option. If it doesn't, we must already have the library ready to go already and don't need this extra library search code.

Lawrence37 replied:
I'm not sure about why the extra logic was needed, but there is a commit for part of the logic for fixing a build issue on openSUSE (https://github.com/artpixls/ART/issues/64). Couldn't hurt to check if this is still an issue. I will see if I can reproduce the original problem and check if this branch is ok.

Lawrence37 replied:
I can't reproduce with openSUSE Tumbleweed, and this branch configures just fine.

Lawrence37 commented on cmake/StandardProjectSettings.cmake:
```diff
@@ -0,0 +1,42 @@
+        set(CMAKE_BUILD_TYPE
+            RelWithDebInfo
+            CACHE STRING "Choose the type of build." FORCE
+        )
```
The default was Debug (and RawPedia says it is Release)

digitalcarp replied:
I'll update it to Release.

Edit: Done

Lawrence37 reviewed (commented):
The LibRaw files are synchronized with the upstream LibRaw with `git subtree`. Will the .gitattributes formatting cause inconvenience?

digitalcarp commented:
I rebased to latest dev. As part of the rebase I changed the first two commits such that `rtengine/libraw/**` is excluded from any line ending changes. This makes `git subtree` work without inconvenience.

Now only 3 files have line ending changes:
- win.cmake
- clean.bat
- tools/buildRT.bat

Lawrence37 commented:
Additional notes for the documentation/changelog:
- The `USE_OLD_CXX_ABI` option has been removed.
- The default build type is "Release", which matches what is claimed in RawPedia. Previously, the default was "Debug".
